parallellism/par.scala

```diff
@@ -0,0 +1,39 @@
+package parallellism.par;
+
+import java.util.concurrent.{ExecutorService, TimeUnit, Callable, Future};
+
+type Par[A] = ExecutorService => Future[A]
+object Par {
+  def run[A](s: ExecutorService)(a: Par[A]) : Future[A] = a(s)
+  def unit[A](value: A): Par[A] = (es: ExecutorService) => UnitFuture(value)
+
+  private case class UnitFuture[A](get: A) extends Future[A] {
+    def isDone = true
+    def get(timeout: Long, units: TimeUnit) : A = get
+    def isCancelled = false
+    def cancel(evenIfRunning: Boolean) : Boolean = false
+  }
+  def fork[A](a: => Par[A]): Par[A] = es => es.submit(new Callable[A] {
+    def call = a(es).get
+  })
+  def lazyUnit[A](value: => A) : Par[A] = fork(unit(value))
+  // how to fix map2 to respect timeout from future?
```

Comment:
wrap in a Future?

Reply:
wrap it with a future would solve the problem as long as we use the delta from the resolution of the first future as the timeout value for the second future.